### ISSUE TITLE
Navigation does not terminate when reaching destination (most houses are no drive-ins)

### ISSUE THREAD

**Issue by sjvudp** [osmandapp/OsmAnd#16594]
# 🐞 routing report

### Routing engine

- [X] OsmAnd's in-app offline routing
- [ ] Any online routing provider (YOURS, openrouteservice, OSRM, etc.)

### Routing Profile

I had been using the car profile

### Actual and expected routes

![Screenshot](https://user-images.githubusercontent.com/6734808/222935734-55d61b12-2512-40e3-a12c-61d51d04f131.jpg)

See the screenshot; it tells everything: I could not get much closer to the destination with the car, but still Osmand wanted me to drive around the block or so.

### Is this a regression?

 Yes, the previous version in which this bug was not present was: ....

## 🌍 Your Environment

**OsmAnd Version:**

<pre><code>

4.3.12
</code></pre>

**Device and Android/iOS version:**
Oppo Finx X5 Lite ColorOS 13
**Maps used (online or offline):**

- [X] Offline maps offered within the OsmAnd app for download.

- [ ] Online (tile / raster) maps 

**Anything else relevant?**
Maybe: The last part is a private road, but for private houses that is not uncommon.
Maybe that fact was confusing Osmand.

**Comment by Zirochkabila** [comment]
@sjvudp Try downloading the night version https://osmand.net/docs/versions/nightly_versions
Or restore the profile settings to the default ones.
If this doesn't help, try reinstalling the app. But, make a backup beforehand


https://user-images.githubusercontent.com/104760013/223460929-30be1568-0cbd-4883-8f07-e789719fbb0c.mp4

**Comment by Zirochkabila** [comment]
OsmAnd~ 4.4.0#14228mqta, Released: 2023-03-10
[logcat (73).log](https://github.com/osmandapp/OsmAnd/files/10951327/logcat.73.log)

In order to display the behavior, we need to set the destination in the house itself, as in the video, then the navigation will not end even if you have arrived at the destination (in fact, you have not arrived, since the destination is in the house, and the car cannot get inside)

https://user-images.githubusercontent.com/104760013/224555671-040d5a1d-166f-4e8b-a8dc-3c16bc51271d.mp4

**Comment by Zirochkabila** [comment]
It is possible to reproduce 4/8 and the algorithm by which it is played back is unclear - random. For now, we will observe

**Comment by sonora** [comment]
@Zirochkabila Does the setting `Navigation settings / Voice prompts /Announcement time` have any impact on the reproducability? Because it looks like we scale not just the announcement timing with this setting, but in the end also the fact when a destination is considered reached:
https://github.com/osmandapp/OsmAnd/blob/92dfd2d842bc38389418f9ab242d340941c6b506/OsmAnd/src/net/osmand/plus/routing/RoutingHelper.java#L536-L539
So if things are sporadic, it could be that that setting is set too ambitious under the circumstances ...

**Comment by Zirochkabila** [comment]
@sonora It's not clear. But, when checking the "4/8", there were no manipulations with the settings

**Comment by sonora** [comment]
What I mean is if you try with "_Announcement time=Early_", is it then more like 0/8?

**Comment by Zirochkabila** [comment]
@sonora I may need to check it

**Comment by sjvudp** [comment]
OK, as I reported the issue, I feel empowered to add some more details, also apologizing for the lack of earlier feedback:
So first for the current Play Store version, here's the suggested route:
![Route-1](https://user-images.githubusercontent.com/6734808/225754826-94c7b25c-2566-4d7e-8a09-df7c29eaf596.jpg)
What you see is that the access road is avoided, and instead a "way around" is suggested.

1. Bug: Interestingly from the endpoint suggested there is no access path to the house, but the "bee line" is a bit shorter than the one when stopping at the main road. The route should end at the main road if the access road is considered "unpassable by car", however. When planning the route there was no message of any kind that the destination is unreachable.

Here are the route options (German). I disabled the one marked.
![Route-Options](https://user-images.githubusercontent.com/6734808/225756308-2ea88d71-b63f-474e-9cc1-f8b4d6ccf880.jpg)

After the change, the route looks OK:
![Route-2](https://user-images.githubusercontent.com/6734808/225756634-d8fbbea5-25bf-4d2c-880f-764d18503221.jpg)

As the problem *might* be caused by incorrect road types in OSM, I checked the properties; here they are:
![OSM](https://user-images.githubusercontent.com/6734808/225756825-e944e985-d452-4d90-8ad9-0c4245f3c939.jpg)

Finally here's the scenery as seen by Google maps:
![Google Maps](https://user-images.githubusercontent.com/6734808/225757940-4bd968aa-fa47-4e70-a1a9-d3c19aad06c9.jpg)

**Comment by sjvudp** [comment]
> In order to display the behavior, we need to set the destination in the house itself, ...

Considering that most navigation will be done using a "street address", the marker will be on the center of the house most likely. So that should be considered to be the "normal case". Also the typical navigation systems handle that correctly.

**Comment by sjvudp** [comment]
The initial problem (as posted) seems to be: Osmand wanted the driver to take a different route, ignoring the fact that the car had arrived (using a different route). Most commercial navigation systems handle that case correctly, however.
Maybe the bug is that Osmand thinks: "You cannot drive where I cannot imagine you to drive" ;-)

**Comment by sonora** [comment]
Ok, so I think it's not a tolerance issue, as I initially suspected .

Instead, I think it is as follows: Whether you have reached your destination is not measrued vs. the actual destination you have set / searched, but vs. the **end point of the route** found and calculated to it. (Otherwise the navigation would never stop in the quite common use cases where your actual destination is e.g. in a garden or building, but no roads extend all the way to it, i.e. your route ends a little distance away.)

In your case now you have discovered a situation like that, where the closest route ends some distance away from your actual destination. But then you have found a way to directly drive to your actual destination. You consider this "arrived", but you are too far away from the end of **OsmAnd's** route end point  to trigger the navigation to finish.

I am  not currently clear how to mitigate this. Potentially by an additional check to immediately end the navigation whenever you are closer to the actual destination than the end of the calculated route would take you. But it could create artifacts, if e.g. you pass your destination on a motorway or on an bridge in rather short distance, you would not want OsmAnd to stop navigating there. Instead take you to the next exit or ramp, then back towards the destination, even if the final route end may be a bit further away from the destination than you already passed it. We have issues complaining that OsmAnd e.g. sometimes routes you to a destination on the other side of a river from where you want to go, that needs to be considered together with solving this issue.

**Comment by Zirochkabila** [comment]
@sonora  I tested your theory "Announcement time=Early" and the result is similar. I got 2/5 with different options (Early, Normal, In the last meters).

**Comment by sonora** [comment]
Yes, the root cause is a different one, see above. Thanks for testing, though!

**Comment by sjvudp** [comment]
Still I wonder about the root cause:

- Is the original road to the house considered "undrivable"? That would explain why OsmAnd insisted to drive elsewhere.
- Doesn't OsmAnd "re-route" when driving a different path as intended/planned? Most typical navigation systems do that. If OsmAnd would have done too, shouldn't it detect "I have no idea how you got there, but you arrived" (or similar). That would be better than insisting on driving around the block and go back by foot (actually it does not suggest the latter, but Google maps does a similar thing).
